feat(eslint-plugin): [no-duplicate-enum-values] add rule

aifreedom opened the pull request:
<!--
👋 Hi, thanks for sending a PR to typescript-eslint! 💖
Please fill out all fields below and make sure each item is true and [x] checked.
Otherwise we may not be able to review your PR.
-->

## PR Checklist

-   [x] Addresses an existing open issue: fixes #2693
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

<!-- Description of what is changed and how the code change does that. -->

This PR creates a new rule in `eslint-plugin` to disallow duplicate enum values. Per discussion in the issue (#2693), in this version, it only checks enum members initialized with literals, and does not cover the ones without an initializer or initialized with an expression.

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-enum-values.ts:
```diff
@@ -0,0 +1,73 @@
+    },
+    hasSuggestions: true,
+    messages: {
+      duplicateMember: 'Duplicate enum member with value {{value}}.',
```
Hmm, this phrasing makes me think that the member is what's duplicated -- like having two members with the same keys. How about...

```suggestion
      duplicateValue: 'Duplicate enum member value {{value}}.',
```

JoshuaKGoldberg commented on packages/eslint-plugin/tests/rules/no-duplicate-enum-values.test.ts:
```diff
@@ -0,0 +1,123 @@
+        },
+      ],
+    },
+  ],
```
There should probably be test cases for falsy values: `''` and `0`. The code seems to correctly check for `=== undefined` instead of general truthiness now, but you never know what'll change in the future...

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-enum-values.ts:
```diff
@@ -0,0 +1,73 @@
+      return (
+        node.type === AST_NODE_TYPES.Literal && typeof node.value === 'number'
+      );
+    }
```
[Nitpicking] Is there a need to separate these two functions out? They both check that `node.type === AST_NODE_TYPES.Literal`.

I'd suggest either making a single separate function:

```ts
function getExpressionLiteralValue(node: TSESTree.Expression) {
  if (node.type === AST_NODE_TYPES.Literal) {
    switch (typeof node.value) {
      // ...
    }  
  }
}

...or inlining them below:

```ts
let value: string | number | undefined;

if (node.type === AST_NODE_TYPES.Literal) {
  // ...
}
```

Just suggestions, feel free to ignore if you don't like either of these changes 😄

JoshuaKGoldberg commented on packages/eslint-plugin/src/rules/no-duplicate-enum-values.ts:
```diff
@@ -0,0 +1,73 @@
+    docs: {
+      description: 'Disallow duplicate enum member values',
+      recommended: false,
+      suggestion: true,
```
This rule doesn't make suggestions in its reports.

```suggestion
```

JoshuaKGoldberg reviewed (changes requested):
The new rule and tests generally looks great, very nicely done @aifreedom! 

Requesting changes for added test coverage. I can't view the Codecov coverage report at the moment to see why the file only has 80% 🙃 -- will take another look when more tests are added.

aifreedom reviewed (commented):
The coverage drop is a bit weird. It's an unrelated rule's coverage that dropped (`no-magic-numbers`). The coverage for the new rule is 100%.

![image](https://user-images.githubusercontent.com/70373/164882375-ba1190a1-638c-4f16-9de2-97f4000a7cb6.png)

aifreedom commented:
Oh, no… I updated the branch using the “rebase” button and it seemed to have forced push the commits. Sorry about the inconvenience. @JoshuaKGoldberg you can review the last two commits which have the updates addressing your comments.

JoshuaKGoldberg commented:
Ha, no worries - thanks for the heads up!

And yeah I don't know what's going on with Codecov. The service has been flaky off and on the last few... years. 🙃

JoshuaKGoldberg reviewed (approved):
Fantastic, thanks so much @aifreedom!
